File: examples/.net/core/ui.case.cs
using System.Text.Json;
using AppProtocol.Example.DotNet.Core.Shared;
using Microsoft.AspNetCore.Components;

namespace AppProtocol.Example.DotNet.Core;

public sealed class UiContext : AppBaseContext
{
    public object? Renderer { get; init; }
    public object? Router { get; init; }
    public object? Store { get; init; }
    public IAppHttpClient? Api { get; init; }
    public IDictionary<string, object?>? Packages { get; init; }
    public IDictionary<string, object?>? Extra { get; init; }
}

public abstract class BaseUiCase<TState>
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        WriteIndented = true,
    };

    protected BaseUiCase(UiContext context, TState initialState)
    {
        Ctx = context;
        State = initialState;
    }

    protected UiContext Ctx { get; }
    protected TState State { get; private set; }

    public abstract RenderFragment View();

    public virtual Task TestAsync() => Task.CompletedTask;

    protected void SetState(TState nextState)
    {
        State = nextState;
    }

    protected static TMaterialized Materialize<TMaterialized>(object? value)
    {
        if (value is TMaterialized typed)
        {
            return typed;
        }

        var json = JsonSerializer.Serialize(value, JsonOptions);
        var materialized = JsonSerializer.Deserialize<TMaterialized>(json, JsonOptions);
        if (materialized is null)
        {
            throw new InvalidOperationException($"Failed to materialize {typeof(TMaterialized).Name}.");
        }

        return materialized;
    }
}

File: examples/flutter/core/ui.case.dart
import 'shared/app_base_context.dart';
import 'shared/app_infra_contracts.dart';
import 'domain.case.dart';

class UiContext extends AppBaseContext {
  const UiContext({
    required super.correlationId,
    required super.logger,
    super.executionId,
    super.tenantId,
    super.userId,
    super.config,
    this.renderer,
    this.router,
    this.store,
    this.api,
    this.packages,
    this.extra,
  });

  final dynamic renderer;
  final dynamic router;
  final dynamic store;
  final AppHttpClient? api;
  final Dict? packages;
  final Dict? extra;
}

typedef UIState = Map<String, dynamic>;

abstract class BaseUiCase<TState extends UIState> {
  BaseUiCase(this.ctx, this.state);

  final UiContext ctx;
  TState state;

  Object view();

  Future<void> test() async {}

  void setState(Map<String, dynamic> partial) {
    state = {...state, ...partial} as TState;
  }
}
